rct_optimizations/test/noise_qualification_utest.cpp

```diff
@@ -0,0 +1,468 @@
+  EXPECT_TRUE(abs(output.j.mean) - abs(q_ver.y()) < 1.0e-14);
+  EXPECT_TRUE(abs(output.k.mean) - abs(q_ver.z()) < 1.0e-14);
+  EXPECT_TRUE(abs(output.w.mean) - abs(q_ver.w()) < 1.0e-14);
+}
+
+TEST(NoiseTest, 2DNoiseTest)
+{
+  //make target
+  test::Target target(4, 4, 0.025);
+
+  //Load camera intrinsics
+  //camera is a kinect
+  auto camera = test::makeKinectCamera();
+
+  //reserve observations
+  std::size_t obs_cnt = 35;
+  std::vector<PnPProblem> perturbed_problem_set;
+  perturbed_problem_set.reserve(obs_cnt);
+
+  Eigen::Isometry3d target_loc = Eigen::Isometry3d::Identity();
+  Eigen::Isometry3d camera_loc = Eigen::Isometry3d::Identity();
+
+  camera_loc.translate(Eigen::Vector3d(0.0,0.0,1.0));
+  camera_loc.rotate(Eigen::AngleAxisd(M_PI,Eigen::Vector3d::UnitX()));
+
+  Eigen::Quaterniond q_ver;
+  Eigen::Matrix3d m = camera_loc.rotation();
+  q_ver = m;
+
+  //add noise boilerplate
+  const double mean = 0.0;
+  const double stddev = 0.001;
+  std::random_device rd{};
+  std::mt19937 generator{rd()};
+  std::normal_distribution<double> dist(mean, stddev);
+
+  //create observations
+  for (std::size_t i = 0; i < obs_cnt; ++i)
+  {
+    PnPProblem instance;
+    //guess inital position
+    instance.camera_to_target_guess= camera_loc;
+    instance.intr = camera.intr;
+
+    Correspondence2D3D::Set corr;
+    EXPECT_NO_THROW
+    (
+    corr = getCorrespondences(camera_loc,
+                              target_loc,
+                              camera,
+                              target,
+                              true);
+   );
+
+   instance.correspondences = corr;
+
+   for (std::size_t j = 0;  j < instance.correspondences.size(); ++j)
+   {
+     double wobblex = dist(generator);
+     double wobbley = dist(generator);
+     instance.correspondences[j].in_image(0) += wobblex;
+     instance.correspondences[j].in_image(1) += wobbley;
+   }
+
+   perturbed_problem_set.push_back(instance);
+  }
+
+ PnPNoiseStat output = rct_optimizations::qualifyNoise2D(perturbed_problem_set);
+
+  EXPECT_TRUE(output.x.std_dev < 1.5 * stddev);
+  EXPECT_TRUE(output.y.std_dev < 1.5 * stddev);
+  EXPECT_TRUE(output.z.std_dev < 1.5 * stddev);
+  EXPECT_TRUE(output.i.std_dev < 1.5 * stddev);
+  EXPECT_TRUE(output.j.std_dev < 1.5 * stddev);
+  EXPECT_TRUE(output.k.std_dev < 1.5 * stddev);
+  EXPECT_TRUE(output.w.std_dev < 1.5 * stddev);
+
+  //Absolute value of quaternion is taken, since quaternions equal their oppoisite
+  EXPECT_TRUE(abs(output.x.mean - camera_loc.translation()(0)) < 1.5 * stddev);
+  EXPECT_TRUE(abs(output.y.mean - camera_loc.translation()(1)) < 1.5 * stddev);
+  EXPECT_TRUE(abs(output.z.mean - camera_loc.translation()(2)) < 1.5 * stddev);
+  EXPECT_TRUE(abs(output.i.mean) - abs(q_ver.x()) < 1.5 * stddev);
+  EXPECT_TRUE(abs(output.j.mean) - abs(q_ver.y()) < 1.5 * stddev);
+  EXPECT_TRUE(abs(output.k.mean) - abs(q_ver.z()) < 1.5 * stddev);
+  EXPECT_TRUE(abs(output.w.mean) - abs(q_ver.w()) < 1.5 * stddev);
```

Comment:
I think we can compare position a little bit more accurately. Using the camera projection matrix we can come up with the differentials of (x, y, z) with respect to our perturbed input (u, v)
- delta_x < (z / f_x) * delta_u
  - `dx < (camera_loc.translation.z() / camera.fx()) * stddev`
- delta_y < (z / f_y) * delta_u
  - `dy < (camera_loc.translation.z() / camera.fy()) * stddev`
- delta_z < (f_x * x * (u - c_x)^-2) * delta_u
  - `dz < (camera.fx() * camera_loc.translation.x() * std::pow((u - camera.cx()), -2.0)) * stddev`
  - `u` in this case is obtained by projecting the actual point into the camera

As far as orientation, I don't think we can directly compare it to the pixel noise level (i.e. `stddev`), so I would suggest setting a somewhat arbitrary value as the threshold and comparing distance between the mean and expected quaternion with:

```c++
EXPECT_LT(Eigen::Quaterniond(output.w.mean, output.i.mean, output.j.mean, output.k.mean).angularDistance(q_ver), angular_threshold);
```